### PR TITLE
Added support for P2G

### PR DESCRIPTION
**Pull request by virmht**
Added support for P2G. 

TODO:

Tests for P2G have to be corrected. Also, calculation of PER and Accuracy is currently incorrect. This needs to be corrected as well.

### CODE REVIEW

**Comment by virmht** [comment]
Hello! I need help in calculating the Phone Error Rate and Accuracy. Also, I need the expected output for test inputs. Can someone review my code ?

**Comment by nshmyrev** [comment]
Hi.  The pull request looks great, thank you. Few things I'd like to change in overall approach

1) Please make input dictionary for training for p2g the same format as for g2p. There is no need in additional format, just parse the inputs and outputs differently

2) Please do not introduce a separate p2g model, you can use the same model, you just need to change the inputs

3) The unit test fails, please check why

**Comment by virmht** [comment]
@nshmyrev can you tell me the formula for calculating Phone Error Rate using edit distance ?

**Comment by nshmyrev** [comment]
Check phonetisaurus calculate-ER script. You first need to align with Levenshtein and then count insertions, deletions and subs.

https://github.com/andrenatal/phonetisaurus-emscripten/blob/master/phonetisaurus/script/phonetisaurus-calculateER

**Comment by virmht** [comment]
Thank you!

**Comment by nshmyrev** [comment]
Hey, the pull request does not look right, look for the changes, it diffs every line. You added windows newlines everywhere, its better to remove them.

File: g2p_seq2seq/g2p.py [hunk 1]
```diff
@@ -80,10 +81,14 @@ def load_decode_model(self):
                                                             "vocab.grapheme"))
     self.ph_vocab = data_utils.load_vocabulary(os.path.join(self.model_dir,
                                                             "vocab.phoneme"))
-
-    self.rev_ph_vocab =\
-      data_utils.load_vocabulary(os.path.join(self.model_dir, "vocab.phoneme"),
-                                 reverse=True)
+    if(self.mode and self.mode == 'g2p'):
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Python does not require parenthesis and does not require checks like C. Also watch the spaces.

File: g2p_seq2seq/g2p.py [hunk 2]
```diff
@@ -139,10 +144,16 @@ def prepare_data(self, train_path, valid_path, test_path):
                                 test_path)
     # Read data into buckets and compute their sizes.
     print ("Reading development and training data.")
-    self.valid_set = self.__put_into_buckets(valid_gr_ids, valid_ph_ids)
-    self.train_set = self.__put_into_buckets(train_gr_ids, train_ph_ids)
+    if(self.mode and self.mode == 'g2p'):
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Also here

File: g2p_seq2seq/g2p.py [hunk 3]
```diff
@@ -364,18 +423,48 @@ def interactive(self):
...
         break
-      print(self.decode_word(word))
-
+      if(self.mode and self.mode == 'g2p'):
+        print(self.decode_word(word))
+      else:
+        print(self.decode_pronunciation(word.strip().split()))
+  
+  
+  def calc_edit_distance(self, s1, s2):
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Keep phoneme error rate feature in a separate pull request please.

File: g2p_seq2seq/g2p.py [hunk 4]
```diff
@@ -364,18 +423,48 @@ def interactive(self):
         break
       if not word:
         break
-      print(self.decode_word(word))
-
+      if(self.mode and self.mode == 'g2p'):
+        print(self.decode_word(word))
+      else:
+        print(self.decode_pronunciation(word.strip().split()))
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: No need to strip before split

File: g2p_seq2seq/g2p.py (hunk 3 above)
Comment (virmht) [reply g2p_seq2seq/g2p.py]: I will keep PER in a separate PR. For this PR, should I not calculate PER at all ?

**Comment by virmht** [comment]
@nshmyrev , Please review changes. I have removed PER and resolved your comments. Please feel free to suggest edits, if any.

File: g2p_seq2seq/g2p.py (hunk 1 above)
Comment (virmht) [reply g2p_seq2seq/g2p.py]: Done!

File: g2p_seq2seq/g2p.py (hunk 2 above)
Comment (virmht) [reply g2p_seq2seq/g2p.py]: Done!

File: g2p_seq2seq/g2p.py (hunk 4 above)
Comment (virmht) [reply g2p_seq2seq/g2p.py]: Done!

File: g2p_seq2seq/g2p.py (hunk 3 above)
Comment (virmht) [reply g2p_seq2seq/g2p.py]: Done!

File: g2p_seq2seq/g2p.py [hunk 5]
```diff
@@ -414,7 +478,10 @@ def decode(self, decode_lines, output_file=None):
     if output_file:
       for word in decode_lines:
         word = word.strip()
-        phonemes = self.decode_word(word)
+        if self.mode and self.mode == 'g2p':
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Like I wrote, it is not C, you can just use

```
if self.mode == 'g2p':
```

and that will be enough. self.mode will never be NULL like in C.

Comment (virmht) [reply g2p_seq2seq/g2p.py]: Done!

File: tests/g2p_unittest.py [hunk 6]
```diff
@@ -9,7 +9,7 @@ class TestG2P(unittest.TestCase):
   def test_train(self):
     model_dir = "tests/models/train"
     with g2p.tf.Graph().as_default():
-      g2p_model = g2p.G2PModel(model_dir)
```
[the remark this answers is not in the document]
Comment (virmht) [reply tests/g2p_unittest.py]: Done!

**Comment by virmht** [comment]
Hey @nshmyrev , please have a look at the changes now.

**Comment by nshmyrev** [comment]
Looks great, thank you very much, cool start!

**Comment by virmht** [comment]
Thank you. Please check PR #97 as well.
